category_theory/discrete_category.lean

```diff
@@ -0,0 +1,77 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+
+import data.ulift
+import category_theory.natural_transformation
+import category_theory.isomorphism
+import category_theory.functor_category
+
+namespace category_theory
+
+universes u₁ v₁ u₂ v₂
+
+def discrete (α : Type u₁) := α
+
+@[extensionality] lemma plift.ext {P : Prop} (a b : plift P) : a = b :=
```

Comment:
"Orphaning" an extensionality lemma like this is a bad idea. It means the behavior of `ext` will change depending on whether this module `category_theory.discrete_category` is imported, and especially since imports are transitive in Lean, this is something one has little control over.

The best place to put this would be either together with the definition of `plift` (impossible in this case), or with the definition of `ext`, or in some module which we can assume is "always" imported--I believe `logic.basic` qualifies. Also possible is to define the lemma here and give it the `extensionality` attribute locally.

However, I don't really understand why we are writing extensionality lemmas for subsingletons. Isn't this something we can solve once and for all by marking `subsingleton.elim` with `extensionality`? Or by modifying `ext` to try applying it?

Reply:
I've moved this next to `ulift.ext` in `tactics/ext.lean`.

`ext` doesn't want to use `subsingleton.elim` for now. I agree this would be a good addition, but I think requires modifying `ext`.

Reply:
Just pinging @cipher1024 in case he knows readily how hard this would be.